### PR TITLE
Explore: Show ANSI colors when highlighting matched words in the logs panel

### PR DESCRIPTION
**Pull request by oliverfrye**
**What this PR does / why we need it**:

Currently, ANSI color codes are rendered in the Explore view for Loki logs. However, if a line filter is used (e.g. `|= "info"`) then the ANSI color codes are no longer rendered, and are instead displayed in their escaped form (e.g. `[36m`) in the log line. This is because a `Highlighter` component is used to highlight the searched for word in the output, which is not ANSI-aware.

image

This PR modifies the existing ANSI-aware `LogMessageAnsi` component to be capable of highlighting the searched for word (still using the `Highlighter` component) in addition to the ANSI coloring.

image

**Which issue(s) this PR fixes**:

Fixes #40265 

**Special notes for your reviewer**:

This is my first Grafana PR, and first time working with React. I'm not sure if my use of multiple `Highlighter` components is problematic in terms of performance - my intention was to avoid having to re-implement its functionality inside `LogMessageAnsi`.

A known issue: highlighting won't work across chunks (blocks of text separated by ANSI escape characters, as I understand) because the `Highlighter` component wraps each `chunk` individually. I don't think this is actually a problem in practice, however, because Loki doesn't support matching line filters that have substrings containing escape codes (see https://github.com/grafana/loki/issues/2861 for more details)

### CODE REVIEW

**Comment by gabor** [comment]
hi @oliverfrye , thank you for the pull request. could you please sign the CLA as mentioned in https://github.com/grafana/grafana/pull/40971#issuecomment-952595122 ? thanks.

**Comment by oliverfrye** [comment]
> hi @oliverfrye , thank you for the pull request. could you please sign the CLA as mentioned in [#40971 (comment)](https://github.com/grafana/grafana/pull/40971#issuecomment-952595122) ? thanks.

Hi @gabor, my apologies for the delay. I'm still awaiting review from my employer's legal team before I am permitted to sign the CLA. I will work on having it signed as soon as possible.

**Comment by gabor** [comment]
> > hi @oliverfrye , thank you for the pull request. could you please sign the CLA as mentioned in [#40971 (comment)](https://github.com/grafana/grafana/pull/40971#issuecomment-952595122) ? thanks.
> 
> Hi @gabor, my apologies for the delay. I'm still awaiting review from my employer's legal team before I am permitted to sign the CLA. I will work on having it signed as soon as possible.

hi @oliverfrye ,sure, no problem.  👍

**Comment by oliverfrye** [comment]
Hi @gabor, I got approval from my company, so the CLA is now signed and this is ready for review. Thank you for your patience 🙏

**Comment by gabor** [comment]
hi @oliverfrye , the PR looks good, it is working fine, good job!

the only thing i am still investigating is backward-compatibility... you see, we export `LogMessageAnsi` from `@grafana/ui`, so any change to it's `Props` can be a backward-incompatible change to `@grafana/ui` , so we have to change it's props very carefully...

currently those extra props (highlight related) basically re-export props from `react-highlight-words`. this can be problematic, you see, if later we decide to upgrade to a new version of `react-highlight-words`, it might cause a backward-incompatibility in `LogMessageAnsi`.

for this reason, i'd prefer to modify `LogMessageAnsi`'s Props to be roughly like this:
```typescript
interface Props {
  value: string;
  highlight?: {
    searchWords: string[];
    highlightClassName: string;
  },
};
```

we store all highlight-related variables in it's own structure, and we make the structure optional, so this will be backward-compatible with old code.

also, we do not have to send in the `findChunks` function, we can just import it from `@grafana/data`, the way `LogRowMessage.tsx` does it. this is important, because that function has a very large type-signature, and this way we are not publishing that in the public API.

there are two ways to go from here:
1. you could do  this change in this PR. or,
2. we can merge your PR as it is, in it's current form, and i will fix the Props API in a separate pull request.

which one do you prefer?

**Comment by oliverfrye** [comment]
@gabor that's excellent feedback, thank you! Your suggestions make it much cleaner.

I've implemented the requested changes and squashed / rebased my PR to include them.

**Comment by gabor** [comment]
hi @oliverfrye , thanks for the changes. it seems the typechecking failed on your changes, the reason is that you need a `@ts-ignore` above the react-highlight import, like it is at https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/components/Logs/LogRowMessage.tsx#L7-L8

(this ts-ignore solution is not optimal, but that's what the other places in grafana-ui do too, so for now let's go with that 😄 )

**Comment by oliverfrye** [comment]
@gabor ah, what a shame. The type error didn't show up in my error, which was unexpected. Nevertheless, I've added `@ts-ignore` like it is in `LogRowMessage.tsx`.

File: packages/grafana-ui/src/components/Logs/LogRowMessage.tsx
Comment (gabor) [inline packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:68]: i think we should use the `needsHighlighter` boolean in the ansi-highlight case too, to be in sync with the other branches, something like this:

```suggestion
    const highlight = needsHighlighter ? {searchWords, highlightClassName} : undefined;
    return <LogMessageAnsi value={entry} highlight={highlight} />;
```

**Review by gabor — CHANGES_REQUESTED** [review]
nice work 👍 ! it works fine, and the code is basically good too, except one minor issue i described below...

we should only allow ansi-highlighting when `needsHighlighter` is `true`, there are some checks in that boolean, like max-string-length etc.

**Comment by oliverfrye** [comment]
Good call, thanks 👍

**Review by gabor — APPROVED** [review]
works nicely!

File: packages/grafana-ui/src/components/Logs/LogRowMessage.tsx
Comment (gabor) [inline packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:68]: i'm sorry, the change i requested made the linter angry 😢 , i think we need to put spaces inside the curly-braces... to be sure, please run `yarn lint`, then fix the problem and run `yarn lint` again to be sure the problems is solved. thanks!

**Review by gabor — CHANGES_REQUESTED** [review]
sorry, there is a linter error that needs to be fixed before we merge this. otherwise everything is ok.

File: packages/grafana-ui/src/components/Logs/LogRowMessage.tsx
Comment (oliverfrye) [reply packages/grafana-ui/src/components/Logs/LogRowMessage.tsx]: OK, no problem, should be sorted now.

**Review by gabor — APPROVED** [review]
LGTM'

**Comment by gabor** [comment]
@oliverfrye thanks for the pull-request, it has been merged!

**Pull request merged**
